File: rgraph.py
#!/usr/bin/env python3

from mpi4py import MPI
from dataset_io import *

import numpy as np
import math
import sys
import getopt
import json
from datetime import datetime
from metricspace import DistPointVector, DistGraph

comm = MPI.COMM_WORLD
myrank = comm.Get_rank()
nprocs = comm.Get_size()

outfile=None
method="cvor" # or gvor or bf or ct
cover=1.55
leaf_size=10
num_centers=10
verbosity=1
stats = {}
stats_file=None

tree_assignment="multiway"
query_balancing="static"
queries_per_tree=-1

def usage():
    global outfile, method, cover, leaf_size, num_centers, tree_assignment, query_balancing, queries_per_tree, stats_file
    if myrank == 0:
        sys.stderr.write(f"Usage: {sys.argv[0]} [options] -i <points> -r <radius>\n")
        sys.stderr.write(f"Options: -c FLOAT cover tree base [{cover:.3f}]\n")
        sys.stderr.write(f"         -l INT   leaf size [{leaf_size}]\n")
        sys.stderr.write(f"         -m INT   num centers [{num_centers}]\n")
        sys.stderr.write(f"         -q INT   queries per tree [{queries_per_tree}]\n")
        sys.stderr.write(f"         -v INT   verbosity level [{verbosity}]\n")
        sys.stderr.write(f"         -o FILE  output sparse graph\n")
        sys.stderr.write(f"         -A STR   tree assignment method (one of: static, multiway) [{tree_assignment}]\n")
        sys.stderr.write(f"         -B STR   load balancing method (one of: static, steal) [{query_balancing}]\n")
        sys.stderr.write(f"         -M STR   querying method (one of: gvor, cvor, ct, bf, ctrma) [{method}]\n")
        sys.stderr.write(f"         -j FILE  stats file\n")
        sys.stderr.write(f"         -F       fix total centers\n")
        sys.stderr.write(f"         -h       help message\n")
        sys.stderr.flush()
    sys.exit(1)

if __name__ == "__main__":

    infile = None
    radius = -1

    fix_num_centers = False

    try: opts, args = getopt.getopt(sys.argv[1:], "c:l:m:Fv:o:i:r:q:A:B:M:j:h")
    except getopt.GetoptError as err: usage()

    for o, a in opts:
        if o == "-i": infile = a
        elif o == "-r": radius = float(a)
        elif o == "-c": cover = float(a)
        elif o == "-l": leaf_size = int(a)
        elif o == "-m": num_centers = int(a)
        elif o == "-q": queries_per_tree = int(a)
        elif o == "-v": verbosity = int(a)
        elif o == "-o": outfile = a
        elif o == "-A": tree_assignment = a
        elif o == "-B": query_balancing = a
        elif o == "-F": fix_num_centers = True
        elif o == "-j": stats_file = a
        elif o == "-M": method = a
        elif o == "-h": usage()
        else: assert False, "unhandled option"

    if not fix_num_centers: num_centers *= nprocs

    if infile is None or radius < 0: usage()

    if not tree_assignment in ("static", "multiway"):
        sys.stderr.write(f"error: '{tree_assignment}' is an invalid tree assignment method!\n")
        sys.stdout.flush()
        sys.exit(1)

    if not query_balancing in ("static", "steal"):
        sys.stderr.write(f"error: '{query_balancing}' is an invalid query balancing method!\n")
        sys.stdout.flush()
        sys.exit(1)

    if not method in ("ct", "bf", "gvor", "cvor", "ctrma"):
        sys.stderr.write(f"error: '{method}' is an invalid query method!\n")
        sys.stdout.flush()
        sys.exit(1)

    comm.barrier()
    t = -MPI.Wtime()
    points = DistPointVector(infile, comm)
    t += MPI.Wtime()
    maxtime = comm.reduce(t, op=MPI.MAX, root=0)

    num_vertices = points.totsize()

    if verbosity >= 1:
        if myrank == 0: sys.stdout.write(f"[v1,time={maxtime:.3f}] Read file '{infile}' [size={num_vertices},dim={points.num_dimensions()}]\n")
        sys.stdout.flush()

    parameters = {}
    parameters["infile"] = infile
    parameters["method"] = method
    parameters["cover"] = cover
    parameters["leaf_size"] = leaf_size
    parameters["num_centers"] = num_centers
    parameters["tree_assignment"] = tree_assignment
    parameters["query_balancing"] = query_balancing
    parameters["queries_per_tree"] = queries_per_tree

    stats["parameters"] = parameters

    t = -MPI.Wtime()

    if method == "bf": graph = points.brute_force_systolic(radius, verbosity)
    elif method == "ct": graph = points.cover_tree_systolic(radius, cover, leaf_size, verbosity)
    elif method == "gvor": graph = points.ghost_tree_voronoi(radius, cover, leaf_size, num_centers, tree_assignment, query_balancing, queries_per_tree, verbosity)
    elif method == "cvor": graph = points.cover_tree_voronoi(radius, cover, leaf_size, num_centers, tree_assignment, query_balancing, queries_per_tree, verbosity)
    elif method == "ctrma": graph = points.cover_tree_rma(radius, cover, leaf_size, verbosity)

    t += MPI.Wtime()
    maxtime = comm.reduce(t, op=MPI.MAX, root=0)

    dist_comps = comm.gather(points.dist_comps(), root=0)
    comm_times = comm.gather(points.my_comm_time(), root=0)
    comp_times = comm.gather(points.my_comp_time(), root=0)
    idle_times = comm.gather(points.my_idle_time(), root=0)

    stats["dist_comps"] = dist_comps
    stats["comm_times"] = comm_times
    stats["comp_times"] = comp_times
    stats["idle_times"] = idle_times
    stats["runtime"] = maxtime

    if method == "gvor" and query_balancing == "steal":
        stats["steal_attempts"] = comm.gather(points.steal_attempts(), root=0)
        stats["steal_successes"] = comm.gather(points.steal_successes(), root=0)
        stats["steal_services"] = comm.gather(points.steal_services(), root=0)
        stats["my_steal_comp_time"] = comm.gather(points.my_steal_comp_time(), root=0)
        stats["my_steal_time"] = comm.gather(points.my_steal_time(), root=0)
        stats["my_poll_time"] = comm.gather(points.my_poll_time(), root=0)
        stats["my_response_time"] = comm.gather(points.my_response_time(), root=0)
        stats["my_allreduce_time"] = comm.gather(points.my_allreduce_time(), root=0)

    #if myrank == 0:
    #    for i in range(nprocs):
    #        sys.stdout.write(f"[rank={i},dist_comps={format_large_number(dist_comps[i])},comp_time={comp_times[i]:.3f},comm_time={comm_times[i]:.3f},idle_time={idle_times[i]:.3f}]\n")

    sys.stdout.flush()
    comm.barrier()

    edges = graph.num_edges(num_vertices,0)
    density = edges/num_vertices

    stats["num_points"] = num_vertices
    stats["num_edges"] = edges
    stats["num_procs"] = nprocs

    if myrank == 0:
        if method == "ct":
            sys.stdout.write(f"[v0,time={maxtime:.3f},p={nprocs}] found neighbors [v={num_vertices},e={edges},e/v={density:.3f},d={format_large_number(sum(dist_comps))},c={cover:.2f},l={leaf_size},M={method}]\n")
        elif method == "ctrma":
            sys.stdout.write(f"[v0,time={maxtime:.3f},p={nprocs}] found neighbors [v={num_vertices},e={edges},e/v={density:.3f},d={format_large_number(sum(dist_comps))},c={cover:.2f},l={leaf_size},M={method}]\n")
        elif method == "bf":
            sys.stdout.write(f"[v0,time={maxtime:.3f},p={nprocs}] found neighbors [v={num_vertices},e={edges},e/v={density:.3f},d={format_large_number(sum(dist_comps))},M={method}]\n")
        elif method == "gvor":
            sys.stdout.write(f"[v0,time={maxtime:.3f},p={nprocs}] found neighbors [v={num_vertices},e={edges},e/v={density:.3f},d={format_large_number(sum(dist_comps))},c={cover:.2f},l={leaf_size},m={num_centers},M={method},A={tree_assignment},B={query_balancing},q={queries_per_tree}]\n")
        elif method == "cvor":
            sys.stdout.write(f"[v0,time={maxtime:.3f},p={nprocs}] found neighbors [v={num_vertices},e={edges},e/v={density:.3f},d={format_large_number(sum(dist_comps))},c={cover:.2f},l={leaf_size},m={num_centers},M={method},A={tree_assignment},B={query_balancing},q={queries_per_tree}]\n")

    sys.stdout.flush()


    if outfile:

        t = -MPI.Wtime()
        graph.write_edge_file(num_vertices, outfile)
        t += MPI.Wtime()
        maxtime = comm.reduce(t, op=MPI.MAX, root=0)

        if verbosity >= 1:
            if myrank == 0:
                sys.stdout.write(f"[v1,time={maxtime:.3f}] wrote edges to file '{outfile}'\n")
            sys.stdout.flush()

    if myrank == 0 and stats_file:
        with open(stats_file, "w") as f:
            json.dump(stats, f, indent=4)

    sys.exit(0)

File: rgraph.cpp
#include <mpi.h>
#include <stdio.h>
#include <iostream>
#include <numeric>
#include <string>
#include <string.h>
#include <unistd.h>

#include "utils.h"
#include "point_vector.h"
#include "dist_point_vector.h"
#include "dist_graph.h"
#include "timer.h"

MPI_Comm comm;
int myrank, nprocs;

Real radius = -1;
const char *infile = NULL;

Real cover = 1.55;
Index leaf_size = 10;
Index queries_per_tree = -1;
Index num_centers = 25;
int verbosity = 1;

const char *outfile = NULL;
const char *tree_assignment = "multiway";
const char *query_balancing = "static";
const char *method = "ct";

void parse_cmdline(int argc, char *argv[]);
int main_mpi(int argc, char *argv[]);
int main(int argc, char *argv[])
{
    MPI_Init(&argc, &argv);
    MPI_Comm_dup(MPI_COMM_WORLD, &comm);
    MPI_Comm_rank(comm, &myrank);
    MPI_Comm_size(comm, &nprocs);
    parse_cmdline(argc, argv);
    int err = main_mpi(argc, argv);
    MPI_Comm_free(&comm);
    MPI_Finalize();
    return err;
}

int main_mpi(int argc, char *argv[])
{
    Timer timer(comm);

    timer.start();
    DistPointVector points(infile, comm);
    timer.stop();
    timer.wait();

    Index num_vertices = points.gettotsize();

    if (verbosity >= 1)
    {
        if (!myrank) printf("[v1,%s] Read file '%s' [size=%lld,dim=%d]\n", timer.repr().c_str(), infile, num_vertices, points.num_dimensions());
        fflush(stdout);
    }

    DistGraph graph(comm);

    MPI_Barrier(comm);
    timer.start();
    if      (!strcmp(method, "bf")) points.brute_force_systolic(radius, graph, verbosity);
    else if (!strcmp(method, "ct")) points.cover_tree_systolic(radius, cover, leaf_size, graph, verbosity);
    else if (!strcmp(method, "gvor")) points.ghost_tree_voronoi(radius, cover, leaf_size, num_centers, tree_assignment, query_balancing, queries_per_tree, graph, verbosity);
    else if (!strcmp(method, "cvor")) points.cover_tree_voronoi(radius, cover, leaf_size, num_centers, tree_assignment, query_balancing, queries_per_tree, graph, verbosity);
    else if (!strcmp(method, "ctrma")) points.cover_tree_rma(radius, cover, leaf_size, graph, verbosity);
    timer.stop();
    timer.wait();

    Index edges = graph.num_edges(num_vertices);
    Real density = (edges+0.0)/num_vertices;

    if (!myrank) printf("[v0,%s] found neighbors [vertices=%lld,edges=%lld,density=%.3f]\n", timer.repr().c_str(), num_vertices, edges, density);
    fflush(stdout);

    if (outfile)
    {
        MPI_Barrier(comm);
        timer.start();
        graph.write_edge_file(num_vertices, outfile);
        timer.stop();
        timer.wait();

        if (verbosity >= 1)
        {
            if (!myrank) printf("[v1,%s] wrote edges to file '%s'\n", timer.repr().c_str(), outfile);
            fflush(stdout);
        }
    }

    return 0;
}

void parse_cmdline(int argc, char *argv[])
{
    auto usage = [&](int err, bool print)
    {
        if (print)
        {
            fprintf(stderr, "Usage: %s [options] -i <points> -r <radius>\n", argv[0]);
            fprintf(stderr, "Options: -c FLOAT cover tree base [%.2f]\n", cover);
            fprintf(stderr, "         -l INT   leaf size [%lld]\n", leaf_size);
            fprintf(stderr, "         -m INT   num centers [%lld]\n", num_centers);
            fprintf(stderr, "         -q INT   queries per tree [%lld]\n", queries_per_tree);
            fprintf(stderr, "         -v INT   verbosity level [%d]\n", verbosity);
            fprintf(stderr, "         -o FILE  output sparse graph\n");
            fprintf(stderr, "         -A STR   tree assignment method (one of: static, multiway) [%s]\n", tree_assignment);
            fprintf(stderr, "         -B STR   load balancing method (one of: static, steal) [%s]\n", query_balancing);
            fprintf(stderr, "         -M STR   querying method (one of: gvor, cvor, ct, bf, ctrma) [%s]\n", method);
            fprintf(stderr, "         -F       fix total centers\n");
            fprintf(stderr, "         -h       help message\n");
        }

        MPI_Finalize();
        std::exit(err);
    };

    bool fix_num_centers = false;

    int c;
    while ((c = getopt(argc, argv, "c:l:m:Fv:o:i:r:q:A:B:M:h")) >= 0)
    {

        if      (c == 'i') infile = optarg;
        else if (c == 'r') radius = atof(optarg);
        else if (c == 'c') cover = atof(optarg);
        else if (c == 'l') leaf_size = atoi(optarg);
        else if (c == 'm') num_centers = atoi(optarg);
        else if (c == 'q') queries_per_tree = atoi(optarg);
        else if (c == 'v') verbosity = atoi(optarg);
        else if (c == 'o') outfile = optarg;
        else if (c == 'A') tree_assignment = optarg;
        else if (c == 'B') query_balancing = optarg;
        else if (c == 'F') fix_num_centers = true;
        else if (c == 'M') method = optarg;
        else if (c == 'h') usage(0, myrank == 0);
    }

    if (!fix_num_centers) num_centers *= nprocs;

    if (!infile || radius < 0) usage(1, myrank == 0);

    if (strcmp(tree_assignment, "static") && strcmp(tree_assignment, "multiway"))
    {
        if (!myrank) fprintf(stderr, "error: '%s' is an invalid tree assignment method!\n", tree_assignment);
        MPI_Finalize();
        std::exit(1);
    }

    if (strcmp(query_balancing, "static") && strcmp(query_balancing, "steal"))
    {
        if (!myrank) fprintf(stderr, "error: '%s' is an invalid query balancing method!\n", query_balancing);
        MPI_Finalize();
        std::exit(1);
    }

    if (strcmp(method, "ct") && strcmp(method, "bf") && strcmp(method, "gvor") && strcmp(method, "cvor") && strcmp(method, "ctrma"))
    {
        if (!myrank) fprintf(stderr, "error: '%s' is an invalid querying method!\n", method);
        MPI_Finalize();
        std::exit(1);
    }
}
